### DISCUSSION TITLE
Pyright cannot resolve 3rd party libraries imports - Neovim

### DISCUSSION THREAD

**Discussion by alanmicah** [microsoft/pyright#4341] [General]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
In Neovim when I attempt to import a 3rd party library such as NumPy I get an error 'Pyright:  Import "NumPy" cannot be resolved'.
I have attempted this some other 3rd party libraries and the same thing is happening. Built in libraries don't seem to produce errors. [1]

However Numpy I am still able to get LSP suggested completions and function definitions. I have shown an example image below with `numpy.array()` [2]. But with other 3rd party libraries, for example, cv2 (OpenCV) I am shown the LSP suggested completion for importing but I do not get LSP completions or definitions when trying to use cv2 functions. [3]

Another example is with PyTorch I get import cannot be resolved errors however, I still am able to get LSP suggested completions and definitions for PyTorch functions, example image shown below [4].

bs4 says "Import bs4 could not be resolved from source", however LSP suggested completions and definitions still work.

In Neovim printing `print(sys.version)` reveals: 3.10.8 | packaged by conda-forge | (main, Nov 22 2022, 08:25:29) [Clang 14.0.6 ]

Pylance works in VSCode completely fine with no issues.

**To Reproduce**
Using within an activated Conda environment
- macOS 12.6.1
- iTerm2 3.4.16
- Neovim v0.8.0
- pyright 1.1.284 (installed globably via NPM) 
-  lua files: `use("neovim/nvim-lspconfig")` and `require("lspconfig").pyright.setup(config())`

Conda/pip and example libraries
- conda 22.9.0 (miniconda)
- pip 22.3.1
- Python 3.10.8 (/Users/{user}/opt/miniconda3/bin/python)
- opencv-python 4.6.0.66
- numpy 1.23.4
- torch                     1.14.0.dev20221031          pypi_0    pypi
- torchvision               0.15.0.dev20221031          pypi_0    pypi

**Expected behavior**
To not display the error and be able to import and use installed 3rd party libraries via pip and/or conda.

**Screenshots or Code**
Import not resolved issue [1]
<img width="475" alt="Screenshot 2022-12-14 at 22 02 08" src="https://user-images.githubusercontent.com/31675641/207724910-3cee1a1a-9ee9-42ff-b8be-a6f071bfae42.png">
---

Completion with definitions working for `import numpy` [2]
<img width="883" alt="Screenshot 2022-12-14 at 22 29 29" src="https://user-images.githubusercontent.com/31675641/207729594-f810ebe9-b890-493a-8e03-2787dc7da62d.png">
---

Errors with 3rd party library imports and functions. [3]
<img width="703" alt="Screenshot 2022-12-14 at 22 22 23" src="https://user-images.githubusercontent.com/31675641/207729690-fba36403-511d-426f-8ab0-4771ec0d867b.png">
---

PyTorch examples [4]
<img width="745" alt="Screenshot 2022-12-14 at 22 46 52" src="https://user-images.githubusercontent.com/31675641/207731998-852ed4cf-262b-4ed2-94b3-966d10c40be5.png">

<img width="947" alt="Screenshot 2022-12-14 at 22 44 46" src="https://user-images.githubusercontent.com/31675641/207731974-56da37f8-9ef7-4722-a808-260825d175df.png">
---

I'm quite new to Neovim and apologise for any ignorance I may have.

**Comment by erictraut** [comment]
This is a configuration issue, not a bug. Moving to a discussion topic.

**Reply by alanmicah** [reply]
My bad. Thank you!

**Reply by erictraut** [reply]
You need to tell pyright about your Python environment. There are several ways to do this.

VS Code provides a convenient control for selecting the environment from among all of the environments found on your machine. Neovim doesn't have this mechanism, so you'll need to use one of the others.

You can use the "python.pythonPath" setting to specify the path to the interpreter you are using. See [this documentation](https://github.com/microsoft/pyright/blob/main/docs/settings.md) for details about settings.

You can also set the Python environment in your shell prior to launching neovim. Pyright will use a shell exec of `python3` to determine which Python environment is configured. 

To debug your configuration, you can look at pyright's output logs. I'm not sure how to do that from within Neovim. In VS Code, there is an "Output" tab, and you can select "Pyright" to view the logs.

**Reply by alanmicah** [reply]
Hi erictraut, thanks for your reply.

Is this what you mean. So in my .zshrc I have set:
```
__conda_setup="$('/Users/{user}/opt/miniconda3/bin/conda' 'shell.zsh' 'hook' 2> /dev/null)"
if [ $? -eq 0 ]; then
    eval "$__conda_setup"
else
    if [ -f "/Users/{user}/opt/miniconda3/etc/profile.d/conda.sh" ]; then
        . "/Users/{user}/opt/miniconda3/etc/profile.d/conda.sh"
    else
        export PATH="/Users/{user}/opt/miniconda3/bin:$PATH"
    fi
fi
unset __conda_setup
```
 `which python` points to the miniconda python, `python --version` seems to point to the same version.

Within Neovim if I execute `python print(sys.version)` it also prints out the python version in conda environment.

In Neovim for :`LspLog` I get a bunch of these messages:
<img width="996" alt="Screenshot 2022-12-14 at 23 26 39" src="https://user-images.githubusercontent.com/31675641/207737145-c07fe912-34ca-4952-b623-bc43f1aae433.png">
 this:

**Reply by erictraut** [reply]
I know next to nothing about neovim, so I probably won't be of much help here. You may need to consult with the maintainers of the [neovim pyright LSP plugin](https://github.com/fannheyward/coc-pyright) to get assistance from them.

**Reply by alanmicah** [reply]
Ahh haha no worries. Will do.
On a side Pyright works wonderfully in VSCode and is so useful, thank you!

**Reply by turboslapper** [reply]
Hey man did you ever figure it out? Having the same issue.

**Reply by happyTonakai** [reply]
Same issue here... Did you solve it?

**Reply by Keshav-writes-code** [reply]
same issue here too
